src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,299 @@
+	if unaggregated.satisfies == partiallySatisfiesRange {
+		if completedAttrs.Retention <= unaggregated.retention {
+			// If the longest aggregated cluster for some reason has lower retention
+			// than the unaggregated cluster then we prefer the unaggregated cluster
+			// as it has a complete data set and is always the most granular.
+			result[0] = unaggregated.clusterNamespace
+			completedAttrs = unaggregated.clusterNamespace.Options().Attributes()
+		}
+	}
+
+	// Take any partially aggregated namespaces with longer retention or
+	// same retention with more granular resolution that may contain
+	// a matching metric.
+	for _, n := range r.partialAggregated {
+		if n.Options().Attributes().Retention > completedAttrs.Retention {
+			// Higher retention.
+			result = append(result, n)
+		} else if n.Options().Attributes().Retention == completedAttrs.Retention &&
+			n.Options().Attributes().Resolution < completedAttrs.Resolution {
+			// Same retention but more granular resolution.
+			result = append(result, n)
+		}
+	}
+
+	return namespaceCoversPartialQueryRange, result, nil
+}
+
+type reusedAggregatedNamespaceSlices struct {
+	completeAggregated []ClusterNamespace
+	partialAggregated  []ClusterNamespace
+}
+
+func (slices reusedAggregatedNamespaceSlices) reset(
+	size int,
+) reusedAggregatedNamespaceSlices {
+	// Initialize arrays if yet uninitialized.
+	if slices.completeAggregated == nil {
+		slices.completeAggregated = make([]ClusterNamespace, 0, size)
+	} else {
+		slices.completeAggregated = slices.completeAggregated[:0]
+	}
+
+	if slices.partialAggregated == nil {
+		slices.partialAggregated = make([]ClusterNamespace, 0, size)
+	} else {
+		slices.partialAggregated = slices.partialAggregated[:0]
+	}
+
+	return slices
+}
+
+// aggregatedNamespaces filters out clusters that do not meet the filter
+// condition, and organizes remaining clusters in two lists if possible.
+//
+// NB: If fanout aggregation is disabled, no clusters will be returned as either
+// partial or complete candidates. If fanout aggregation is forced to enabled
+// then no filter is applied, and all namespaces are considered viable. In this
+// case, the filter is used to determine if returned namespaces have the
+// complete set of metrics.
+//
+// NB: If fanout optimization is enabled, add any aggregated namespaces that
+// have a complete set of metrics to the completeAggregated slice list. If this
+// optimization is disabled, or if none of the aggregated namespaces are
+// guaranteed to have a complete set of all metrics, they are added to the
+// partialAggregated list.
+func aggregatedNamespaces(
+	all ClusterNamespaces,
+	slices reusedAggregatedNamespaceSlices,
+	filter func(ClusterNamespace) bool,
+	opts *storage.FanoutOptions,
+) reusedAggregatedNamespaceSlices {
+	// Reset reused slices.
+	slices = slices.reset(len(all))
+
+	// Force disable fanout to any aggregated namespaces.
+	if opts.FanoutAggregated == storage.FanoutForceDisable {
```

Comment:
As discussed before, if `FanoutAggregated` is actually disabled, then we can just probably return immediately after checking the unaggregated namespace. (i.e. return unaggregated as long as that is default or enabled) and then don't need to check it again at all.

Reply:
Yeah, it gets to the same conclusion but with more steps; will try and short circuit it instead